File: micronaut-jdbc/src/main/java/com/github/frederikpietzko/repository/VisitRepository.java
package com.github.frederikpietzko.repository;

import com.github.frederikpietzko.dto.PetDto;
import com.github.frederikpietzko.dto.VisitDto;
import com.github.frederikpietzko.model.Type;
import com.github.frederikpietzko.model.VisitEntity;
import io.micronaut.data.annotation.Query;
import io.micronaut.data.jdbc.annotation.JdbcRepository;
import io.micronaut.data.model.query.builder.sql.Dialect;
import io.micronaut.data.repository.CrudRepository;
import io.micronaut.serde.annotation.Serdeable;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Optional;

@JdbcRepository(dialect = Dialect.POSTGRES)
public interface VisitRepository extends CrudRepository<VisitEntity, Long> {

  @Query("""
    SELECT v.id as visit_id, v.description, v.date_time, v.pet_id, p.name as pet_name, p.age as pet_age, p.type as pet_type
    FROM visit v
    JOIN pet p ON v.pet_id = p.id
    """)
  List<VisitProjection> findAllWithPet();

  @Query("""
    SELECT v.id as visit_id, v.description, v.date_time, v.pet_id, p.name as pet_name, p.age as pet_age, p.type as pet_type
    FROM visit v
    JOIN pet p ON v.pet_id = p.id
    WHERE v.id = :id
    """)
  Optional<VisitProjection> findByIdWithPet(Long id);

  @Serdeable
  record VisitProjection(
    Long visit_id,
    String description,
    OffsetDateTime date_time,
    Long pet_id,
    String pet_name,
    int pet_age,
    Type pet_type
  ) {
    public VisitDto toDto() {
      return new VisitDto(
        visit_id,
        description,
        new PetDto(pet_id, pet_name, pet_age, pet_type),
        date_time
      );
    }
  }
}

File: ktor/src/main/kotlin/com/github/frederikpietzko/VisitRepository.kt
package com.github.frederikpietzko

import com.github.frederikpietzko.model.Pet
import com.github.frederikpietzko.model.Visit
import org.jetbrains.exposed.sql.JoinType
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.insertAndGetId
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.transactions.experimental.newSuspendedTransaction

object VisitRepository {
  suspend fun getVisits() = newSuspendedTransaction {
    VisitTable
      .join(PetTable, JoinType.INNER)
      .selectAll()
      .map { it.toVisit() }
  }

  suspend fun findVisitById(id: Long) = newSuspendedTransaction {
    VisitTable
      .join(PetTable, JoinType.INNER)
      .selectAll()
      .where { VisitTable.id eq id }
      .map { it.toVisit() }
      .singleOrNull()
  }

  suspend fun save(visit: Visit) = newSuspendedTransaction {
    val petId = visit.pet.id ?: PetTable.insertAndGetId {
      it[name] = visit.pet.name
      it[age] = visit.pet.age
      it[type] = visit.pet.type
    }.value

    val visitId = VisitTable.insertAndGetId {
      it[description] = visit.description
      it[pet] = petId
      it[dateTime] = visit.dateTime
    }.value

    visit.copy(id = visitId, pet = visit.pet.copy(id = petId))
  }

  private fun ResultRow.toVisit() = Visit(
    id = this[VisitTable.id].value,
    description = this[VisitTable.description],
    pet = Pet(
      id = this[PetTable.id].value,
      name = this[PetTable.name],
      age = this[PetTable.age],
      type = this[PetTable.type],
    ),
    dateTime = this[VisitTable.dateTime],
  )
}
